src/engine/positionscratchcontroller.h

```diff
@@ -6,18 +6,42 @@
 #include "audio/frame.h"
 
 class ControlObject;
-class RateIIFilter;
-class VelocityController;
+class ControlProxy;
+
+class VelocityController {
+  public:
+    VelocityController();
+
+    void setPD(double p, double d);
+    void reset(double last_error);
+    double observation(double error);
+
+  private:
+    double m_last_error;
+    double m_p, m_d;
+};
+
+class RateIIFilter {
+  public:
+    RateIIFilter();
+
+    void setFactor(double factor);
+    void reset(double last_rate);
+    double filter(double rate);
+
+  private:
+    double m_factor;
+    double m_last_rate;
+};
```

Comment:
I know you just moved this here into the header, but unless you intend to use these two classes outside of `PositionScratchController`, they don't make much sense in the header.

Reply:
I wasn't able to use std::unique_ptr with the definition in ratecontrol.cpp, foward declaration didn't work out.
Any hints how to make this possible are welcome.

And I had the idea to clone `PositionScratchController` --> `WheelScratchCotroller`, so we have one for the waveforms and another one for controller wheels (maybe with some more customization, eg. controls for the max. throw/spinback time, filter parameters and maybe sample interval)

Reply:
> I wasn't able to use std::unique_ptr with the definition in ratecontrol.cpp, foward declaration didn't work out.

I'm gonna have to make an educated blind guess here without a compiler log, but the issue is probably that the ctor and/or dtor got auto-instantiated in the class that used it in the header. In that case, you'll need to declare the respective special member function (ctor/dtor) (in `PositionScratchController`) and then implement it in the `.cpp` (preferably just using an explicit `= default`).

Reply:
I reverted that change and got
`/usr/include/c++/13/bits/unique_ptr.h:97:23: error: invalid application of ‘sizeof’ to incomplete type ‘VelocityController’`

So yes, adding the dtor that fixed it, thanks!
(no I understand the note in [BaseTrackCache::~BaseTrackCache()](https://github.com/mixxxdj/mixxx/blob/f0ce199b4383595ec46c92f974996cdefd198d52/src/library/basetrackcache.cpp#L37-L41))

Reply:
`= default;` didn't suffice btw

Reply:
It should've if you've done it in the `.cpp` (not in the header!)

Anyways, not important in this case